cluster-autoscaler/simulator/delta_cluster_snapshot.go

```diff
@@ -0,0 +1,523 @@
+		deletedNodeInfos: make(map[string]bool),
+	}
+}
+
+// Convenience method to avoid writing loop for adding nodes.
+func (data *internalDeltaSnapshotData) addNodes(nodes []*apiv1.Node) error {
+	for _, node := range nodes {
+		if err := data.addNode(node); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func (data *internalDeltaSnapshotData) addNode(node *apiv1.Node) error {
+	nodeInfo := schedulernodeinfo.NewNodeInfo()
+	if err := nodeInfo.SetNode(node); err != nil {
+		return fmt.Errorf("cannot set node in NodeInfo: %v", err)
+	}
+	return data.addNodeInfo(nodeInfo)
+}
+
+func (data *internalDeltaSnapshotData) addNodeInfo(nodeInfo *schedulernodeinfo.NodeInfo) error {
+	if _, found := data.nodeInfoMap[nodeInfo.Node().Name]; found {
+		return fmt.Errorf("node %s already in snapshot", nodeInfo.Node().Name)
+	}
+	data.nodeInfoMap[nodeInfo.Node().Name] = nodeInfo
+	if data.nodeInfoList != nil {
+		data.nodeInfoList = append(data.nodeInfoList, nodeInfo)
+	}
+	if data.completeNodeInfoList != nil {
+		data.completeNodeInfoList = append(data.completeNodeInfoList, nodeInfo)
+	}
+	return nil
+}
+
+func (data *internalDeltaSnapshotData) clearCaches() {
+	data.nodeInfoList = nil
+	data.completeNodeInfoList = nil
+	data.clearPodCaches()
+}
+
+func (data *internalDeltaSnapshotData) clearPodCaches() {
+	data.podList = nil
+	data.rawPods = nil
+	data.rawPodsTotal = 0
+	data.havePodsWithAffinity = nil
+}
+
+func (data *internalDeltaSnapshotData) updateNode(node *schedulernodeinfo.NodeInfo) error {
+	if _, found := data.nodeInfoMap[node.Node().Name]; found {
+		data.removeNode(node.Node().Name)
+	}
+
+	return data.addNodeInfo(node)
+}
+
+func (data *internalDeltaSnapshotData) removeNode(nodeName string) error {
+	_, found := data.nodeInfoMap[nodeName]
+	if found {
+		// If node was added or modified within this delta, delete this change.
+		delete(data.nodeInfoMap, nodeName)
+	}
+
+	ni, err := data.baseData.GetNodeInfo(nodeName)
+	if err == nil && ni != nil {
+		// If node was found in the underlying data, mark it as deleted in delta.
+		data.deletedNodeInfos[nodeName] = true
+	}
+
+	if err != nil && !found {
+		// Node not found in the chain.
+		return err
+	}
+
+	// Maybe consider deleting from the lists instead. Maybe not.
+	data.clearCaches()
+	return nil
+}
+
+func (data *internalDeltaSnapshotData) addPod(pod *apiv1.Pod, nodeName string) error {
+	if _, found := data.nodeInfoMap[nodeName]; !found {
+		if ni, err := data.baseData.GetNodeInfo(nodeName); err != nil {
+			return err
+		} else {
+			data.nodeInfoMap[nodeName] = ni.Clone()
+		}
+	}
+
+	data.nodeInfoMap[nodeName].AddPod(pod)
+	if data.podList != nil || data.havePodsWithAffinity != nil {
+		data.clearPodCaches()
+	}
+	return nil
+}
+
+func (data *internalDeltaSnapshotData) getNodeForPod(namespace, name string) (string, error) {
+	if data == nil {
+		return "", fmt.Errorf("pod %s/%s not in snapshot", namespace, name)
+	}
+
+	for nodeName, nodeInfo := range data.nodeInfoMap {
```

Comment:
This is costly. And we are doing quite a bit of those in scaledown.
Maybe we should change the interface of RemovePod so it also takes node name?
Acutally maciek suggested that.

Reply:
If we know the name of the node at the time of removing the pod, it would be perfect.

I tried to implement it differently (store pod -> node mapping), but it was even more costly.